core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (C) 2016-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.persistence.cassandra.query
+
+import java.util.UUID
+import java.util.concurrent.ThreadLocalRandom
+
+import akka.annotation.InternalApi
+import akka.persistence.PersistentRepr
+import akka.persistence.cassandra._
+import akka.persistence.cassandra.journal.{ BucketSize, CassandraJournal, TimeBucket }
+import akka.persistence.cassandra.query.EventsByTagStage._
+import akka.serialization.SerializationExtension
+import akka.stream.stage._
+import akka.stream.{ ActorMaterializer, Attributes, Outlet, SourceShape }
+import com.datastax.driver.core._
+
+import scala.annotation.tailrec
+import scala.concurrent.duration._
+import scala.concurrent.{ ExecutionContext, ExecutionContextExecutor, Future }
+import scala.util.{ Failure, Success, Try }
+import java.lang.{ Long => JLong }
+
+import akka.cluster.pubsub.{ DistributedPubSub, DistributedPubSubMediator }
+import akka.persistence.cassandra.query.QueryActorPublisher.Continue
+import akka.persistence.cassandra.query.scaladsl.CassandraReadJournal.CombinedEventsByTagStmts
+import com.datastax.driver.core.utils.UUIDs
+
+/**
+ * Walks the tag_views table.
+ *
+ * For current queries:
+ * Moves to the next timebucket when there are no more events. If
+ * the current time bucket is today then the query ends when it is exhausted.
+ *
+ * For live queries:
```

Comment:
Could it simplify this code if current vs. live behaviour were split into separate stage classes? That would have the added benefit that the live (real-time) stream could be put behind a `BroadcastHub`. That way, _all_ currently open, caught up to real-time, queries for the same tag could be handled with 1 poll.